test/src/org/exist/test/TransactionTestDSL.java

```diff
@@ -0,0 +1,779 @@
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+package org.exist.test;
+
+import com.evolvedbinary.j8fu.Either;
+import com.evolvedbinary.j8fu.function.QuadFunction7E;
+import com.evolvedbinary.j8fu.tuple.Tuple2;
+import org.exist.EXistException;
+import org.exist.collections.Collection;
+import org.exist.collections.triggers.TriggerException;
+import org.exist.dom.persistent.DocumentImpl;
+import org.exist.dom.persistent.NewArrayNodeSet;
+import org.exist.dom.persistent.NodeProxy;
+import org.exist.dom.persistent.NodeSet;
+import org.exist.security.PermissionDeniedException;
+import org.exist.storage.BrokerPool;
+import org.exist.storage.DBBroker;
+import org.exist.storage.lock.Lock;
+import org.exist.storage.txn.Txn;
+import org.exist.test.TransactionTestDSL.BiTransactionScheduleBuilder.BiTransactionScheduleBuilderOperation;
+import org.exist.test.TransactionTestDSL.TransactionScheduleBuilder.NilOperation;
+import org.exist.util.LockException;
+import org.exist.xmldb.XmldbURI;
+import org.exist.xquery.XPathException;
+import org.exist.xquery.XQuery;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.*;
+import java.util.function.Function;
+
+import static com.evolvedbinary.j8fu.Either.Left;
+import static com.evolvedbinary.j8fu.Either.Right;
+import static org.exist.dom.persistent.DocumentImpl.XML_FILE;
+
+/**
+ * A DSL for describing a schedule of
+ * transaction operations upon the database.
+ *
+ * A type-safe builder pattern is provided
+ * for constructing the schedule. Once
+ * the schedule is build a scheduler
+ * can execute it upon the database
+ * and return the results.
+ *
+ * The DSL uses recursive types
+ * in a similar way to a typed heterogeneous
+ * list (such as Shapeless's HList) to ensure
+ * that the each operation in the schedule
+ * receives the correct input type, i.e.
+ * the output type of the previous operation.
+ * At the cost of complexity in implementing
+ * the DSL, the recursive typing makes use of
+ * the DSL by the user much simpler and safer.
+ *
+ * The recursive type implementation was
+ * inspired by {@see https://apocalisp.wordpress.com/2008/10/23/heterogeneous-lists-and-the-limits-of-the-java-type-system/}.
+ *
+ * Example usage for creating a schedule of
+ * two transactions, where each will execute in
+ * it's own thread but operationally linear
```

Comment:
A minor typo here: "it's" (the contraction of "it is") should be "its" (the possessive form of "it").